spring-cloud-commons/src/main/java/org/springframework/cloud/client/hedger/HedgedRequestsExchangeFilterFunction.java

```diff
@@ -0,0 +1,114 @@
+ *
+ *      https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.springframework.cloud.client.hedger;
+
+
+import java.time.Duration;
+import java.util.List;
+
+import org.apache.commons.logging.Log;
+import org.apache.commons.logging.LogFactory;
+import reactor.core.publisher.Flux;
+import reactor.core.publisher.Mono;
+import reactor.util.function.Tuple2;
+
+import org.springframework.lang.NonNull;
+import org.springframework.lang.Nullable;
+import org.springframework.web.reactive.function.client.ClientRequest;
+import org.springframework.web.reactive.function.client.ClientResponse;
+import org.springframework.web.reactive.function.client.ExchangeFilterFunction;
+import org.springframework.web.reactive.function.client.ExchangeFunction;
+
+/**
+ * WebClient filter function that allows for "hedging" requests that take too long: after the specified delay, if
+ * the original request was not completed yet, it will fire one or more "hedged" requests in hopes that they will
+ * return sooner than the original request.
+ *
+ * This is useful to try to cut down on the long tail latencies of requests. For example, one could set the delay
+ * to the 95th percentile latency of the downstream service. Typically that will result in a significant reduction of
+ * 95th percentile latency in exchange for 2-5% traffic increase. (The actual numbers depend on the distribution of
+ * latencies.)
+ *
+ * {@see http://accelazh.github.io/storage/Tail-Latency-Study} for more background on how hedging works.
+ *
+ * @author Csaba Kos
+ * @author Kevin Binswanger
+ */
+public class HedgedRequestsExchangeFilterFunction implements ExchangeFilterFunction {
+
+	private static final Log log = LogFactory
+			.getLog(HedgedRequestsExchangeFilterFunction.class);
+
+	private HedgingPolicy hedgingPolicy;
+	private List<HedgingMetricsReporter> hedgingMetricsReporterList;
+
+	public HedgedRequestsExchangeFilterFunction(
+			HedgingPolicy hedgingPolicy, List<HedgingMetricsReporter> hedgingMetricsReportersList) {
+		this.hedgingPolicy = hedgingPolicy;
+		this.hedgingMetricsReporterList = hedgingMetricsReportersList;
+	}
+
+	@Override
+	@NonNull
+	public Mono<ClientResponse> filter(@NonNull ClientRequest request, ExchangeFunction next) {
+		Duration delay = hedgingPolicy.getDelayBeforeHedging(request);
+		int numHedges = numberOfHedgedRequestsDelayAware(request, hedgingPolicy, delay);
+		return withSingleMetricsReporting(request, next.exchange(request), null)
+				.mergeWith(
+						Flux.range(1, numHedges)
+								.delayElements(delay)
+								.flatMap(hedgeNumber -> withSingleMetricsReporting(request, next.exchange(request), hedgeNumber)
```

Comment:
So in multiple places we say that we only call report for the actual request that we use (1st to succeed). Refresh my memory how this code does not possibly call if multiple of the hedges succeed?